middleware/kubernetes/handler.go

```diff
@@ -53,38 +49,46 @@ func (k Kubernetes) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.M
 			if (dns.CountLabel(name) - 1) < k.AutoPath.NDots {
 				break
 			}
-			// Search "svc.cluster.local" and "cluster.local"
+			origQName := state.QName()
+			// Search "svc.cluster.local." and "cluster.local."
 			for i := 0; i < 2; i++ {
 				path = strings.Join(dns.SplitDomainName(path)[1:], ".")
 				state = state.NewWithQuestion(strings.Join([]string{name, path}, "."), state.QType())
 				records, extra, _, err = k.routeRequest(zone, state)
 				if !k.IsNameError(err) {
+					records = append(records, NewCNAME(origQName, records[0].Header().Name, records[0].Header().Ttl))
 					break
 				}
 			}
 			if !k.IsNameError(err) {
```

Comment:
This breaks out of the outer loop (`for p!= nil`).
I could label the outer loop, and break out from the inner loop line 60 if you prefer.